packages/react-dom/src/events/DOMPluginEventSystem.js

```diff
@@ -923,7 +923,7 @@ function accumulateEnterLeaveListenersForEvent(
   inCapturePhase: boolean,
 ): void {
   const registrationName = event._reactName;
-  if (registrationName === undefined) {
+  if (registrationName === null) {
```

Comment:
Right now this can never be true:

- `accumulateEnterLeaveListenersForEvent` is called from [`accumulateEnterLeaveTwoPhaseListeners`](https://github.com/facebook/react/blob/5bb80b8b17f1e3722e5f0be9002e432d2b352238/packages/react-dom/src/events/DOMPluginEventSystem.js#L970) 
- `accumulateEnterLeaveTwoPhaseListeners` is called from [`EnterLeaveEventPlugin#extractEvents`](https://github.com/facebook/react/blob/22d16cc15df8e570b79c2825fb495c2b98b40bf5/packages/react-dom/src/events/plugins/EnterLeaveEventPlugin.js#L45) with synthetic events that are created from native`mouseover`, `pointerover`, `mouseout` or `pointerout`
- for these native events we have a statically defined `_reactName` in [`userBlockingPairsForSimpleEventPlugin`](https://github.com/facebook/react/blob/50893dbcb7030499eeecfa4b44539b603b757afd/packages/react-dom/src/events/DOMEventProperties.js#L96)

It might make sense to add an invariant or just cast it to a string with a hint to the current data flow.

Reply:
I don't understand this change.

If it was previously sometimes `undefined`, then this changes behavior.

If it was previously sometimes `null`, then this changes behavior.

If it was neither, then this line can be removed, and the type can be made more strict.

Does that make sense?

Reply:
I think maybe you're saying that the type is more loose but for this particular case, we only get a subtype? Can we codify that in types?

Reply:
> If it was previously sometimes undefined, then this changes behavior.

Not according to the flow types. Maybe there are still some files where SyntheticEvent is used that are uncovered? 

> If it was previously sometimes null, then this changes behavior.

It does change but there weren't any test so we're definitely missing some code coverage. This is what I meant with adding a test for `reactName === null`.

> I think maybe you're saying that the type is more loose but for this particular case, we only get a subtype? Can we codify that in types?

Exactly. I'll try something like `UnknownSyntheticEvent { reactName: null }` and `KnownSyntheticEvent { reactName: string }` and see if we can get rid of some checks with earlier type checks.

Reply:
My impression is that this branch has something to do with the `createEventHandle` API. But I might be wrong. It would be good to look in git for why it was added.

Reply:
> Can we codify that in types?

[We can](https://github.com/facebook/react/pull/19564/commits/e75f16343f147cfae98fe220b3ceb63dd7240598). One thing I noticed though is that `new SyntheticEvent` is not type safe at all. Flow seems to think that a generic object was created and doesn't actually type-check the properties at all: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDKWAdhgBZ4YCWAxgKIBueZAKrgQLxgDeYATngCGtDADkhAWzwAuMAGcM-aiQDmYAL7ooAVxKjqcEkVIUqdJiwwAKAJS9UYMMGBgEBACZGA5BjBD5eWpVY0ERcSkCSkEAOjBHZ1cAAwpqeSSwLzx5MBI4P2pJHBg8aTIwJOIyShoGZjYODIAjHT9YRDdqGBgwEr8dHICgkLBqPww4CqqzWssG-AyhEiwEISwEwQwdfmNU+QBuVC1UF1zJvH5+OH4AGgqSPCRCOGkAMT0DIzsMnCumoSaMCwYAAVgM-JttiRBmBVCwLnQwHAmiC8KIwNh8P4SB5MnBsiRfP5RDohN1gbRKLQANYVPaLQLBEhlDDyVC0IyKMB6eRCKB4OYYOTTGoWeoYdhY7gPJ6mUV1Kx2Q7smBDEzVcwKsgAYVVgV4x1OFyu-EGOTwAA98KI8B52Zy-Lz+YLhXLNYLJVxco91TMxVZdUM7KggA

So right now any call to a function expecting `ReactSyntheticEventType` with `new SyntheticEvent` is not sound.

Reply:
> It would be good to look in git for why it was added.

https://github.com/facebook/react/commit/991c3b8193bd6bcc53c23c79eddc659eb41a4c3a#diff-6b784480f6f3297c42bc2e6e46e6f943L829-R818

[`event.dispatchConfig.registrationName` was `undefined | string`](https://github.com/facebook/react/commit/991c3b8193bd6bcc53c23c79eddc659eb41a4c3a#diff-a98878cd54c12afcf7cd54a3c02c4930L30) while `reactName` is assumed to be `string` (even though it [was nullable](https://github.com/facebook/react/blob/991c3b8193bd6bcc53c23c79eddc659eb41a4c3a/packages/react-dom/src/events/DOMModernPluginEventSystem.js#L943) and [still is](https://github.com/facebook/react/blob/master/packages/react-dom/src/events/DOMPluginEventSystem.js#L1037).